File: Episode09-Lives/Monogame/Player.cs
using Microsoft.Xna.Framework;
using Microsoft.Xna.Framework.Graphics;
using Microsoft.Xna.Framework.Input;
using MonoGame.Extended;

namespace Shmup
{
    internal static class Player
    {
        /// <summary>
        /// Static class for Player as there is only 1
        /// </summary>
        #region Class variables
        public static bool Alive = true;
        public static RectangleF Rectangle;
        public static CircleF Circle;
        public static float Shield = 100f;
        public static int Lives = 3;
        public static bool Hidden = false;
        private static float radius = 0f;
        private static float speed = 10;
        private static Texture2D playerImg;
        private static float scale = 0f;
        private static float hideTimer = 0f;
        #endregion
        #region Public Methods
        public static void CreatePlayer(Texture2D playerimg, float scl, float spd)
        {
            playerImg = playerimg;
            scale = scl;
            speed = spd;
            Rectangle = new RectangleF
            (
                Shared.WIDTH / 2 - (playerImg.Width / 2), 
                Shared.HEIGHT - playerImg.Height * scale - 10,
                playerImg.Width * scale,
                playerImg.Height * scale
            );
            radius = MathHelper.Min(playerImg.Width, playerImg.Height) / 2 * scale;
            Circle = new CircleF(Rectangle.Position, radius);
        }
        public static void Hide()
        {
            Hidden = true;
            hideTimer = 0f;
            Rectangle.X = Shared.WIDTH / 2;
            Rectangle.Y = Shared.HEIGHT + 200;
        }
        public static void Update(KeyboardState keyboardState, float dt)
        {
            // unhide if hidden
            if (Hidden)
            {
                hideTimer += dt;
                if (hideTimer > 2) //  restore to centre after 2 second
                {
                    Hidden = false;
                    Rectangle.X = Shared.WIDTH / 2 - Rectangle.Width / 2;
                    Rectangle.Y = Shared.HEIGHT - Rectangle.Height - 10;
                }
            }
            if (keyboardState.IsKeyDown(Keys.Left) || keyboardState.IsKeyDown(Keys.A))
            {
                Rectangle.X -= speed * dt;
                if (Rectangle.X < 0)
                    Rectangle.X = 0;
            }
            else if (keyboardState.IsKeyDown(Keys.Right) || keyboardState.IsKeyDown(Keys.D))
            {
                Rectangle.X += speed * dt;
                if (Rectangle.X > Shared.WIDTH - Rectangle.Width)
                    Rectangle.X = Shared.WIDTH - Rectangle.Width;
            }
            Circle.Position = Rectangle.Center;
        }
        public static void Draw(SpriteBatch spriteBatch)
        {
            spriteBatch.Draw
            (
                texture: playerImg,
                position: Rectangle.Position,
                sourceRectangle: null,
                color: Color.White,
                rotation: 0.0f,
                origin: Vector2.Zero,
                scale: scale,
                effects: SpriteEffects.None,
                layerDepth: 0f
            );
            if (Shared.Debug)
            {
                spriteBatch.DrawRectangle(Rectangle, Color.Blue);
                spriteBatch.DrawCircle(circle: Circle, sides: 20, color: Color.Red, thickness: 1);
            }
        }
        #endregion
    }
}

File: Episode09-Lives/Love2D/player.lua
local Player = {} 	-- simple table for Player. As there is only one, class not required
local Rectangle = require "lib.Rectangle"
local Circle = require "lib.Circle"
local Shared = require "shared"

function Player.setProperties(sprite, scale, speed)
	Player.sprite = sprite
	Player.scale = scale
	Player.speed = speed
	Player.rect = Rectangle(0, 0, sprite:getWidth() * scale, sprite:getHeight() * scale)
	Player.rect.w = sprite:getWidth() * scale
	Player.rect.h = sprite:getHeight() * scale
	Player.rect.x = Shared.WIDTH / 2 - Player.rect.w / 2
	Player.rect.y = Shared.HEIGHT - Player.rect.h - 10
	Player.radius = math.max(Player.rect.width, Player.rect.height) * 0.8 / 2
	Player.circle = Circle(Player.rect.x + Player.radius, Player.rect.y + Player.radius, Player.radius)
	Player.alive = true
	Player.shield = 100
	Player.lives = 3
	Player.hidden = false
	Player.hideTimer = 0
end	

function Player.update(keyboard, dt)
	-- unhide if hidden
	if Player.hidden then
		Player.hideTimer = Player.hideTimer + dt
		if Player.hideTimer > 2 then -- restore to centre after 2 second
			Player.hidden = false
			Player.rect:setX(Shared.WIDTH / 2)
			Player.rect:setY(Shared.HEIGHT - Player.rect.height - 10)
		end
	end
	
	if keyboard.isDown("left") or keyboard.isDown("a") then
		Player.rect:updatePosition(Player.speed * dt * -1, 0) 	-- move Player left
	end
	if keyboard.isDown("right") or keyboard.isDown("d")then	
		Player.rect:updatePosition(Player.speed * dt, 0) 		-- move Player right
	end
	if Player.rect.x < 0 then							-- check if Player x position is off-screen left side
		Player.rect:setX(0)								-- whoops! change it to 0: left side of screen
	end
	if Player.rect.right > Shared.WIDTH then					-- check if Player x position is off-screen right side (take Player width into account)
		Player.rect:setX(Shared.WIDTH - Player.rect.width)		-- whoops! change it to right side of screen, less the Player width		
	end
	Player.circle:update(Player.rect.x + Player.radius, Player.rect.y + Player.radius)
end

function Player.draw()
	if Player.alive then
		love.graphics.setColor(Shared.WHITE)			-- change colour to white
		love.graphics.draw(Player.sprite, Player.rect.x, Player.rect.y, nil, Player.scale, Player.scale)
		if Shared.debug then
			love.graphics.setColor(Shared.RED)
			love.graphics.rectangle("line", Player.rect.x, Player.rect.y, Player.rect.w, Player.rect.h)
			love.graphics.setColor(Shared.BLUE)
			love.graphics.circle('line', Player.circle.x, Player.circle.y, Player.circle.radius)
			love.graphics.setColor(Shared.WHITE)
		end
	end
end

function Player.hide()
	-- hide the sprite by moving it off-screen
	Player.hidden = true
	Player.hideTimer = 0
	Player.rect:update(Shared.WIDTH / 2, Shared.HEIGHT + 200)
end

return Player   